File: javapatterns/src/main/java/org/iut/nantes/state/distributeur/Distributeur.java
package org.iut.nantes.state.distributeur;

class Distributeur
{
  private Etat etatAttente_;
  private Etat etatPiece_;
  private Etat etatBonbonVendu_;
  private Etat etatVide_;

  private Etat etatCourant_;

  private int stockBonbon_;
  
  public Distributeur( int stock )
  {
    etatAttente_ = new EtatAttente( this );
    etatPiece_ = new EtatPiece( this );
    etatBonbonVendu_ = new EtatBonbonVendu( this );
    etatVide_ = new EtatVide( this );

    etatCourant_ = etatAttente_;
    stockBonbon_ = stock;
  }

  // Transitions
  public void insererPiece()
  {
    etatCourant_.insererPiece();
  }
  
  public void annuler()
  {
    etatCourant_.annuler();
  }
  
  public void tournerPoignee()
  {
    etatCourant_.tournerPoignee();
  }

  // Actions
  public void afficherMessage()
  {
    etatCourant_.afficherMessage();
  }

  public void changerEtatAttente()
  {
    etatCourant_ = etatAttente_;
    afficherMessage();
  }

  public void changerEtatPiece()
  {
    etatCourant_ = etatPiece_;
    afficherMessage();
  }

  public void changerEtatBonbonVendu()
  {
    etatCourant_ = etatBonbonVendu_;
    afficherMessage();
  }

  public void changerEtatVide()
  {
    etatCourant_ = etatVide_;
    afficherMessage();
  }

  public int getStockBonbon()
  {
    return stockBonbon_;
  }

  public void decrementeStock()
  {
    --stockBonbon_;
  }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/state/distributeur/Distributeur.kt
package org.iut.nantes.state.distributeur

import org.iut.nantes.state.distributeur.etats.*

class Distributeur {
    var bonbons: Int = 10

    val etatAttente = Attente(this)
    val etatVide = Vide(this)
    val etatUnePiece = UnePiece(this)
    val etatBonbonVendu = BonbonVendu(this)

    var etat: Etat = etatAttente

    fun message() {
        etat.message()
    }

    fun insererPiece() {
        etat.insererPiece()
    }

    fun tournerPoignee() {
        etat.tournerPoignee()
    }

    fun annuler() {
        etat.annuler()
    }

    fun toEtatAttente() {
        etat = etatAttente
    }

    fun toEtatUnePiece() {
        etat = etatUnePiece
    }

    fun toEtatBonbonVendu() {
        etat = etatBonbonVendu
    }

    fun toEtatVide() {
        etat = etatVide
    }
}